### PR TITLE
Test `busy` and `idle` events

### PR DESCRIPTION
**Pull request by kasperisager**

### CODE REVIEW

File: lib/stream.js [hunk 1]
```diff
@@ -111,7 +111,7 @@ module.exports = class UDXStream extends streamx.Duplex {
 
     socket.streams.add(this)
     this
-      .on('close', () => {
+      .prependListener('close', () => {
```
[the remark this answers is not in the document]
Comment (mafintosh) [reply lib/stream.js:114]: @kasperisager should't we just make this an `_` method and call that from _destroy instead? that has the same affect and is clearer imo

File: lib/socket.js [hunk 2]
```diff
@@ -99,6 +99,19 @@ module.exports = class UDXSocket extends events.EventEmitter {
...
+    if (this.idle) this._onbusy()
+
+    this.streams.add(stream)
+  }
+
+  _removeStream (stream) {
+    this.streams.delete(stream)
+    this._closeMaybe()
+
+    if (this.idle) this._onidle()
```
Comment (kasperisager) [inline lib/socket.js:112]: @mafintosh Should this actually happen before `_closeMaybe()`?

Comment (mafintosh) [reply lib/socket.js:112]: yea, i think that makes sense actually.

File: lib/socket.js [hunk 3]
```diff
@@ -133,7 +147,10 @@ module.exports = class UDXSocket extends events.EventEmitter {
   _closeMaybe () {
     if (this._inited && this.closing && this.idle) {
       binding.udx_napi_socket_close(this._handle)
+      return true
     }
+
+    return false
```
[the remark this answers is not in the document]
Comment (mafintosh) [reply lib/socket.js:153]: (just extra safety and makes it a bit clearer also i think)

**Pull request merged**